src/utils/data-access.ts

```diff
@@ -0,0 +1,148 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+import { PromiseResult } from 'aws-sdk/lib/request';
+import { AWSError } from 'aws-sdk/lib/error';
+import DynamoDB, { DocumentClient, QueryOutput } from 'aws-sdk/clients/dynamodb';
+import { Configurations } from './configuration';
+import * as domain from '../domain';
+import { debug } from './logger';
+
+export class DataAccess {
+  // private static instance: DataAccess;
+
+  private readonly tableName: string = Configurations.getInstance().dynamoTableName;
+
+  private docClient: DocumentClient;
+
+  public constructor() {
+    this.docClient = new DynamoDB.DocumentClient(Configurations.getInstance().dynamoParams);
+  }
+
+  // public static getInstance(): DataAccess {
+  //   if (!DataAccess.instance) {
+  //     DataAccess.instance = new DataAccess();
+  //   }
+
+  //   return DataAccess.instance;
+  // }
+
+  private getById(options: {
+    indexName: string;
+    keyCondition: string;
+    expressionAttributeValues: { [key: string]: string };
+  }): Promise<PromiseResult<QueryOutput, AWSError>> {
+    const params = {
+      TableName: this.tableName,
+      IndexName: options.indexName,
+      KeyConditionExpression: options.keyCondition,
+      ExpressionAttributeValues: options.expressionAttributeValues,
+    };
+    debug('params getbyId', params);
+    return this.docClient.query(params).promise();
+  }
+
+  private async put<T>(
+    payload: T,
+    options?: { conditionalExpression: string; expressionAttributeValues: { [key: string]: string } },
+  ): Promise<PromiseResult<DocumentClient.PutItemOutput, AWSError>> {
+    let query = {
+      TableName: this.tableName,
+      Item: payload,
+      ConditionExpression: undefined,
+      ExpressionAttributeValues: undefined,
+    };
+    if (options) {
+      query = {
+        ...query,
+        ConditionExpression: options.conditionalExpression,
+        ExpressionAttributeValues: options.expressionAttributeValues,
+      };
+    }
+    return this.docClient.put(query).promise();
+  }
+
+  public async upsertTrailerRegistration(
+    trailerRegistration: domain.TrailerRegistration,
+  ): Promise<domain.TrailerRegistration> {
+    await this.put<domain.TrailerRegistration>(trailerRegistration);
+    const result = await this.getByVinOrChassisWithMake(trailerRegistration.vinOrChassisWithMake);
+    if (result && result.length) {
+      return result[0];
+    }
+    return null;
+  }
+
+  public async getByVinOrChassisWithMake(vinOrChassisWithMake: string): Promise<domain.TrailerRegistration[]> {
+    const getParams = {
+      indexName: 'vinOrChassisWithMakeIndex',
```

Comment:
Question for us and possibly future requirements ? Do we want to try to enhance the user experience and return user friendly 'errors' or messages depending on what we find/don't find. 
I believe at the moment we default to Internal Server Error

Reply:
Well, user friendly is enough for 40X errors  e.g. 400, 404, 401 but for 500 we should default to `Internal Server Error` and log the actual error in cloud watch. This abstracts internal failures for end users who might get confused by looking at code/database related error.